### DISCUSSION TITLE
DB Public API and Data Structures

### DISCUSSION THREAD

**Discussion by vdmitriyev** [dmarro89/dare-db#34] [Ideas]
A various databases support a variety of data structures to store and process data. 
This thread should help in discussion of a various data structures that could be implemented within DareDB. 

DareDB makes use of [go-redis-hashtable](https://github.com/dmarro89/go-redis-hashtable) implementation

The current discussion also will include discussion of public API of the database

**Comment by vdmitriyev** [comment]
I would like to suggest a basic data structure called "collection". 

### Overview of Collection Structure

* A collection could contain other basic structures. For now only: key/value pairs
* A collection should have a unique name within a user namespace
* A default collection should be called ```default```
* If no collection was specified in the ```get/set``` inquiry, the ```default``` collection should be used to fetch the value
* A collection should provide an ability to iterate over items inside it
	+ Items within the collection should be somehow "sortable"
    + Because, a collections could be very big, a HTTP interface should provide an ability to get keys out of the collection in a "page-by-page" manner
    + Example of the "page by page": 
        - Let us say we have a collection with 123 items and a single page size is ```10```
        - After accessing the first batch of keys in the collection, the server should return an identifier to second batch (e.g., ```nextPage: "2"```)
* A collection should be able to accept values without "keys", in case no keys provided, a key based on UUIDv7 should be generated for a new value
    + UUIDv7 is a Time-Sortable Identifier for Modern Databases. 
	+ More information on UUIDv7 - https://uuid7.com/
    + **Alternative** to UUIDv7 could be ULID (Universally Unique Lexicographically Sortable Identifier) implemented in Go https://github.com/oklog/ulid
    

Important: For now, it is not clear, whether UUIDv7 is a good choice for hashing algorithm used in the DareDB. This needs to be investigated.


### REST API Interface for Collection Structure

There are couple possibilities to organize with work with collection. Which could be divided into 2 main ones: implicit and explicit. 

* The **implicit** one suggest to use GET, POST, DELETE, PUT to work with collections. 
* The **explicit** one, is to use additional verbs (e.g., create, iterate, remove, check, size, etc.) to describe an actions while working with collection. 

The following API interface draft make use of **explicit**. Please, find below the suggestion to work with collection via API (```http://localhost:2605/``` is used as an example):

* List all collections:
    - GET request: ```http://localhost:2605/collections/list/```
    - Response: JSON with all collections
* Create a new collection:
    - GET request: ```http://localhost:2605/collections/create/:collection```
        - ```collection```: string value
    - Response: HTTP code (200, 404, 505, etc.) + message
* Get size of a collection:
    - GET request: ```http://localhost:2605/collections/size/:collection```
        - ```collection```: string value
    - Response: JSON with size of elements in a collection
* Iterate of a collection:
    - GET request: ```http://localhost:2605/collections/iterate/:collection?page=1&pagesize=1000```
        - ```page```: current page with keys
        - ```pagesize```: number of keys within a page (default: 1000)
    - Response: JSON with keys within requested range
* Delete collection:
    - DELETE request: ```http://localhost:2605/collections/delete/:collection```
    - Response: HTTP code (200, 404, 505, etc.) + message

**Reply by dmarro89** [reply]
Hi @vdmitriyev thanks for the idea.
Could you please provide an example ? I'm wondering how this data structure can be used.
Thanks

**Reply by vdmitriyev** [reply]
@dmarro89 sorry, I had no time previously to describe the API interface for the collections in details. I just updated the original suggestion and have added "REST API Interface for Collection Structure" part

**Reply by dmarro89** [reply]
> @dmarro89 sorry, I had no time previously to describe the API interface for the collections in details. I just updated the original suggestion and have added "REST API Interface for Collection Structure" part

Thanks

**Reply by vdmitriyev** [reply]
@dmarro89 additional clarification:
I have called it a "data structure", but in reality it is more like a "logical" abstraction over hash tables, which to improve the user experience and make it more convenient

**Reply by dmarro89** [reply]
> @dmarro89 additional clarification:
> 
> I have called it a "data structure", but in reality it is more like a "logical" abstraction over hash tables, which to improve the user experience and make it more convenient 

Ok perfect, that's what I was asking to you.
Perfect, I totally agree with your proposal.

**Reply by dmarro89** [reply]
Do we want to create a new issue related to it ?

**Reply by vdmitriyev** [reply]
> Do we want to create a new issue related to it ?

If you are ok with the proposal, than a new issue could be created. In order to avoid duplicated information, let's also link this discussion as well as this particular comment flow in the new issue

**Reply by dmarro89** [reply]
Ok, thanks

**Comment by dmarro89** [comment]
@vdmitriyev I've read your proposal. 
Dare-DB is based on a custom hashtable implementation (go-redis-hashtable) and uses this hashtable to store te key/value pairs.
With the introduction of the "collection" concept, do we want to handle multiple hashtables for each single instance of DareDB -> 1 hashtable = 1 collection ?

**Reply by vdmitriyev** [reply]
@dmarro89 thanks for the question. During the creation of proposal, I hoped, that you could help me on that particular questions, because you are the one who knows how  `go-redis-hashtable` works internally. As for me, I have only an "abstract idea", who this could work. Here you go:

* A `collection` should be represented by struct in Go
* This `collection`-struct, should consist of `map` (or something similar to it), that will contain at least a reference to an instance of `go-redis-hashtable`.
    - It means, that after each HTTP request, `collection`-struct should be checked for the respected `go-redis-hashtable`
        + There is also an alternative to access data without using `map`. But it will require a different hash-handling inside (or before entering) `go-redis-hashtable` , which overcomplicated hash-handling (e.g., just add some pre-defined prefix to each hash, to "assign" each key to a particular collection)
    - This `map` will have pointers pointing to instances of `go-redis-hashtable`, where the actual data should be located
    - This  `collection`-struct should act a singleton 
    - Overall memory footprint of  `collection`-struct will not be that big, because the data should be stored in `go-redis-hashtable`
* Also `collection`-struct could contain additional variables/data to help with managing the database
   - Some sort of `mutex`, to ensure,  that is acts as a singleton and could be operated "thread-safe". Actually, this is not that crucial for now, because `collection`-struct will be mostly used to "read" names of the collections and pointers to `go-redis-hashtable` hashtable  instances and will be rarely update
   - A parameter, that will contain a max size of main memory allowed for a collection (this number should be checked, before the underlying/referenced `go-redis-hashtable` will be increased in side)
   - It could also posses information on "access roles" for each collection

P.S. Sorry for my late response. Had not much time recently.

**Reply by dmarro89** [reply]
ok @vdmitriyev thanks.
Let's create an issue about the implementation.

**Comment by dmarro89** [comment]
Hi @vdmitriyev, here the first PR about the creation of the collection abstraction -> https://github.com/dmarro89/dare-db/pull/47
Take a look when you can.

Thanks

**Reply by vdmitriyev** [reply]
Thank you for tis PR. It looks like you contributed with a lot of code, which is amazing 💪 
I will give it a shot, as long as I will have a bit more free time

**Reply by dmarro89** [reply]
Thank you @vdmitriyev 
Don't worry, there's no rush.
Take your time

**Comment by vdmitriyev** [comment]
@dmarro89 

Answer to the comment in PR: https://github.com/dmarro89/dare-db/pull/47#issuecomment-2424617088

Answering here, because don't want to overload PR. Essential discussion has more to do with architectural decisions and not that much source code related.

Let me try to clarify, why using a `redundant` verbs in API endpoints in this particular case could be beneficial. First of all, API of the database should be consistent. Note that `get` and `set` verbs are already used by the database to define its behavior. Second, having an explicit way of performing an action (e.g., specify what exactly will be "done" directly in a URL) can enhance understandability could be facilitate flexibility in behavior. Implicit encode such behavior as part of a HTTP request have its own advantages, of course. I know about suggestions and best practices to design REST API using HTTP request verbs only, but this forces a potential user to always look up documentation in order to understand, what is really going on. 

Additionally, I wanted to separate management of collections (e.g., create, update, delete, rename) under the /collections/ prefix (note – it is plural). However, the work with a collection, should be using a prefix /collection/ (note – it is singular). So, to be able to work in such fashion, additional verbs to describe actions were required (otherwise one may get confused, as it happened to me in a Python script).

**Reply by dmarro89** [reply]
ok Thanks @vdmitriyev.
I Agree with your vision.
I'll merge the PR and then update the endpoints and the python script as well.

**Reply by vdmitriyev** [reply]
@dmarro89 you are welcome! Thank you for merging your PR. I have open an issue in order not to forget to sync description of API for collection handling with implementation - https://github.com/dmarro89/dare-db/issues/54

**Reply by dmarro89** [reply]
Perfetto, thank you @vdmitriyev

**Comment by vdmitriyev** [comment]
@dmarro89 FYI: I have change the title of this discussion, so we could have everything in one place for now - public API (the one that the users of the database see) and data structure because we mainly discussed public API in this case as well)

**Reply by dmarro89** [reply]
Ok perfect!
